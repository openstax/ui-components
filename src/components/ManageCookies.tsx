import React from 'react';
import { ButtonLink } from "./Button";
import { createGlobalStyle } from "styled-components";

const GlobalStyle = createGlobalStyle`
  .cky-btn-revisit { display: none; }
`;

type ManageCookiesLinkProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  wrapper?: (button: React.ReactElement) => React.ReactElement;
};

// documentation for this at https://www.cookieyes.com/documentation/change-cookie-consent-using-cookieyes/
export const ManageCookiesLink = ({children, className, wrapper, ...props}: ManageCookiesLinkProps) => {
  const inBrowser = typeof window === 'object';
  const [cookieYesLoaded, setCookieYesLoaded] = React.useState(false);

  React.useEffect(() => {
    if (inBrowser && !cookieYesLoaded) {
      const handleCkyLoaded = () => setCookieYesLoaded(true);
      if ('getCkyConsent' in window) {
        handleCkyLoaded();
      } else {
        document.addEventListener('cookieyes_banner_load', handleCkyLoaded);
        return () => document.removeEventListener('cookieyes_banner_load', handleCkyLoaded);
      }
    }
    return;
  }, [inBrowser]);

  if (!inBrowser) { return <><GlobalStyle /></>; }

  const button = <ButtonLink
    className={`cky-banner-element${className ? ` ${className}` : ''}`}
    {...props}
  >{children || 'Manage Cookies'}</ButtonLink>;

  return <>
    <GlobalStyle />
    {cookieYesLoaded
       ? typeof wrapper === 'function'
         ? wrapper(button)
         : button
       : null
    }
  </>;
};
