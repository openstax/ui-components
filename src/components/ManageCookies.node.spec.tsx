/** @jest-environment node */
/** eslint-disable @typescript-eslint/no-explicit-any */
import renderer from 'react-test-renderer';
import { ManageCookiesLink } from "./ManageCookies";

describe('ManageCookies outside a browser', () => {
  it('renders nothing', () => {
    const tree = renderer.create(
      <ManageCookiesLink />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
