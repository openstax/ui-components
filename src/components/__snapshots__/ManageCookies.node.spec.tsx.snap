// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageCookies outside a browser renders nothing 1`] = `null`;
